compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -1843,7 +1843,22 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
               }
             }
 
-            Some(methodDefWithoutUselessVars)
+            /* See: https://github.com/scala-js/scala-js/issues/3953 */
+            val methodDefWithPatchedTypes = {
+              val symParamTypes =
+                if (sym.paramss.isEmpty) Nil
+                else sym.paramss.head.map(p => toIRType(p.tpe))
+
+              assert(symParamTypes.length == params.length,
+                  "symParams and vparams have different lengths")
+
+              val patchedParamTypes = collection.mutable.Map(
+                  (params.map(encodeLocalSym(_).name) zip symParamTypes): _*)
```

Comment:
You can use `methodDefWithoutUselessVars.args` instead of `params` here. That's better because we already have their `LocalName`, hence we don't need to re-encode the local syms.

Moreover, and more importantly, we should filter out the parameters for which the IR type is already correct (i.e., the `(arg: js.ParamDef).ptpe == symParamType)`. If the resulting list is empty, which is the overwhelmingly common case, we should not call `patchTypesOfLocals` at all!